### DISCUSSION TITLE
Seamless integration of LongVar (64-bit integer)

### DISCUSSION THREAD

**Discussion by dimitri-justeau** [chocoteam/choco-solver#834] [Ideas]
I believe that integer variables, based on 32-bits integers, are suited for most usages of choco-solver. Operations on 32-bits integers are also apparently faster in Java. However, I know from a recent experience that 64-bit long variables could also be helpful in some situations.

Ideally, having a LongVar extending IntVar for such particular cases would be perfect. However, this involves some design challenges:

- It would need the implementation of backtrackable longs.
- Propagators implemented for int variables should also work seamlessly with long vars.
- Mixing int vars and long vars in propagator should be seamlessly possible.

In summary, I think that if we can implement this feature as a pure extension, without impacting propagators implementation nor solver performances would be a great asset for choco-solver.

If judged interesting by the community, I am ok to get involved in this work.

**Comment by cprudhom** [comment]
As a first step, which does not fit entirely the need, I propose to add the following API to `IntVar`:
- `removeValue(long value, ICause cause)`
- `instantiateTo(long value, ICause cause)`
- `updateLowerBound(long value, ICause cause)`
- `updateUpperBound(long value, ICause cause)`

which automatically deal with `int` casting (if possible).
This would make possible to deal with large range of values in `IntVar`, for instance very close to `Integer` bounds, without unexpected behaviors. This would make easier to prevent from underflow/overflow cases.
For instance, `PropTimes` can be adapted quite easily.

**Comment by dimitri-justeau** [comment]
There might be some radical changes about the way Java handles primitives, which should include generics over primitives:

https://www.levi9.com/articles/10-ways-java-is-getting-better/#anchor-7

https://openjdk.org/projects/valhalla/

I think that we should stay tuned about this :-)

**Comment by cprudhom** [comment]
This argues for upgrading the JDK to a new one
